### PR TITLE
Explain how Clippy and lints work

### PR DESCRIPTION
**Pull request by phansch**
Also for #2666 

This adds a rough overview of how Clippy hooks into rustc and how different lint types are registered with the PluginRegistry. It doesn't explain everything yet, but should give a good overview, hopefully.

It is partly inspired by the [quick rustfmt tour](https://github.com/rust-lang-nursery/rustfmt/blob/master/Contributing.md#a-quick-tour-of-rustfmt). 

[Rendered](https://github.com/phansch/rust-clippy/blob/8b8242feef44fe0bea678604485c76c2ed95c801/CONTRIBUTING.md#how-clippy-works)

### CODE REVIEW

**Comment by phansch** [comment]
Would be good if someone with more experience has a look at this, as I'm not sure if I used the correct terms everywhere.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -147,6 +147,55 @@ enabled as a plugin:
...
+// ...
+pub mod else_if_without_else;
+// ...
+
+pub fn register_plugins(reg: &mut rustc_plugin::Registry) {
+    // ...
+    reg.register_early_lint_pass(box else_if_without_else::ElseIfWithoutElse);
+    // ...
+
+    reg.register_lint_group("clippy_restriction", vec![
```
Comment (Manishearth) [inline CONTRIBUTING.md]: Worth mentioning that this code is autogenerated

Comment (phansch) [reply CONTRIBUTING.md]: Hm, I'm not sure I understand how that code would be autogenerated. It's a normal call to [`Registry.register_lint_group`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_plugin/registry/struct.Registry.html#method.register_lint_group), no?

Comment (phansch) [reply CONTRIBUTING.md]: oh, *that* kind of autogenerated! I've always been modifying it manually until now.. Makes sense to add it :+1:

**Pull request merged**